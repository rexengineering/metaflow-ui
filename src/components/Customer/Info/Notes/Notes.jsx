import React from "react";
import PropTypes from "prop-types";
import InfoSection from "../InfoSection";
import ActionItem from "../../../ActionItem";
import { ActionItemDeleteType } from "../../../../constants";

function Notes({ notes, className, onAddIconClicked, onNoteDeleted }) {
  return (
    <InfoSection
      className={className}
      onAddIconClicked={onAddIconClicked}
      title="Notes"
    >
      {Array.isArray(notes) &&
        notes.map((note) => (
          <ActionItem
            label={note}
            actionType={ActionItemDeleteType}
            onAction={onNoteDeleted}
          />
        ))}
    </InfoSection>
  );
}

Notes.defaultProps = {
  className: "",
  onAddIconClicked: () => {},
  onNoteDeleted: () => {},
};

Notes.propTypes = {
  notes: PropTypes.arrayOf(PropTypes.string).isRequired,
  className: PropTypes.string,
  onAddIconClicked: PropTypes.func,
  onNoteDeleted: PropTypes.func,
};

export default Notes;
