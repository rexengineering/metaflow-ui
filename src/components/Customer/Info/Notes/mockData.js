export default ["Follow up on home services", "Follow up on sales tips"];
