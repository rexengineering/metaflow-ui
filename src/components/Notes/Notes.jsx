import React, { useCallback, useState, useEffect } from "react";
import PropTypes from "prop-types";
import {
  Typography,
  Button,
  Card,
  CardContent,
  CardActions,
  CardHeader,
  IconButton,
  makeStyles,
} from "@material-ui/core";
import { Formik } from "formik";
import { object as yupObject } from "yup";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faTrashAlt } from "@fortawesome/pro-solid-svg-icons";
import TaskField from "../fields/TaskField";
import { TEXT, validationSchemaMapping } from "../../constants/taskTypes";

const useStyles = makeStyles((theme) => ({
  card: {
    marginBottom: theme.spacing(1),
  },
  notesContainer: {
    overflowY: "auto",
  },
  userNote: {
    whiteSpace: "pre-wrap",
  },
}));

function Notes({ onNotesNumberChangeCB }) {
  const [notes, setNotes] = useState([]);
  const classes = useStyles();
  const initialValues = { newNote: "" };
  const validationSchema = yupObject({
    newNote: validationSchemaMapping[TEXT],
  });

  useEffect(() => {
    if (onNotesNumberChangeCB) {
      onNotesNumberChangeCB(notes?.length);
    }
  }, [onNotesNumberChangeCB, notes]);

  const onTextAreaKeyDown = useCallback((event, handleSubmit) => {
    if (event?.keyCode === 13 && !event?.shiftKey) {
      event?.preventDefault();
      event?.stopPropagation();
      handleSubmit();
    }
  }, []);

  const addNote = useCallback(
    ({ newNote }, { resetForm }) => {
      setNotes((currentNotes) => [
        ...currentNotes,
        { id: Math.random(), message: newNote, date: new Date() },
      ]);
      resetForm();
    },
    [setNotes]
  );

  const removeNote = useCallback((id) => {
    setNotes((currentNotes) => currentNotes.filter((note) => note.id !== id));
  }, []);

  return (
    <>
      <Typography variant="h5" align="center">
        Notes
      </Typography>
      <div className={classes.notesContainer}>
        {Array.isArray(notes) &&
          notes.map(({ id, message, date }) => (
            <Card key={id} className={classes.card} variant="outlined">
              <CardHeader
                action={
                  <IconButton onClick={() => removeNote(id)}>
                    <FontAwesomeIcon icon={faTrashAlt} />
                  </IconButton>
                }
                title={date.toLocaleTimeString()}
              />
              <CardContent>
                <Typography
                  className={classes.userNote}
                  data-testid="note-message"
                >
                  {message}
                </Typography>
              </CardContent>
            </Card>
          ))}
      </div>
      <Formik
        initialValues={initialValues}
        validationSchema={validationSchema}
        onSubmit={addNote}
      >
        {({ handleSubmit }) => (
          <Card onSubmit={handleSubmit} component="form">
            <CardContent>
              <TaskField
                id="newNote"
                type={TEXT}
                multiline
                fullWidth
                label="New Note"
                onKeyDown={(event) => onTextAreaKeyDown(event, handleSubmit)}
              />
            </CardContent>
            <CardActions>
              <Button type="submit">Add</Button>
            </CardActions>
          </Card>
        )}
      </Formik>
    </>
  );
}

Notes.propTypes = {
  onNotesNumberChangeCB: PropTypes.func,
};

Notes.defaultProps = {
  onNotesNumberChangeCB: undefined,
};

export default Notes;
