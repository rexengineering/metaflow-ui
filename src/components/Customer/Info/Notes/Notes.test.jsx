import { render, screen } from "@testing-library/react";
import React from "react";
import notesData from "./mockData";
import Notes from ".";

describe("<Notes />", () => {
  const renderNotesComponent = () => {
    return <Notes notes={notesData} />;
  };

  it("should render", () => {
    const { asFragment } = render(renderNotesComponent());
    const initialRender = asFragment();

    expect(initialRender).not.toBeNull();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(renderNotesComponent());
    const initialRender = asFragment();

    expect(initialRender).toMatchSnapshot();
  });

  it("should render all passed notes", () => {
    render(renderNotesComponent());
    notesData.forEach((note) => {
      const noteNode = screen.getByText(note);
      expect(noteNode).toBeInTheDocument();
    });
  });
});
