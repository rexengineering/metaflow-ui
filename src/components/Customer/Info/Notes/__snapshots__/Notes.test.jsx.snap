// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notes /> should match snapshot 1`] = `
<DocumentFragment>
  <section
    class=""
  >
    <div
      class="makeStyles-header-7"
    >
      <h6
        class="MuiTypography-root makeStyles-label-8 MuiTypography-h6"
      >
        Notes
      </h6>
      <button
        class="MuiButtonBase-root MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-plus-circle fa-w-16 "
            data-icon="plus-circle"
            data-prefix="fal"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M384 250v12c0 6.6-5.4 12-12 12h-98v98c0 6.6-5.4 12-12 12h-12c-6.6 0-12-5.4-12-12v-98h-98c-6.6 0-12-5.4-12-12v-12c0-6.6 5.4-12 12-12h98v-98c0-6.6 5.4-12 12-12h12c6.6 0 12 5.4 12 12v98h98c6.6 0 12 5.4 12 12zm120 6c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zm-32 0c0-119.9-97.3-216-216-216-119.9 0-216 97.3-216 216 0 119.9 97.3 216 216 216 119.9 0 216-97.3 216-216z"
              fill="currentColor"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div>
      <div
        class="makeStyles-container-9"
      >
        <p
          class="MuiTypography-root makeStyles-label-10 MuiTypography-body1"
        >
          Follow up on home services
        </p>
        <div
          class="makeStyles-actionElement-11"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButtonSize-12"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-trash-alt fa-w-14 "
                data-icon="trash-alt"
                data-prefix="fal"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M296 432h16a8 8 0 0 0 8-8V152a8 8 0 0 0-8-8h-16a8 8 0 0 0-8 8v272a8 8 0 0 0 8 8zm-160 0h16a8 8 0 0 0 8-8V152a8 8 0 0 0-8-8h-16a8 8 0 0 0-8 8v272a8 8 0 0 0 8 8zM440 64H336l-33.6-44.8A48 48 0 0 0 264 0h-80a48 48 0 0 0-38.4 19.2L112 64H8a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h24v368a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V96h24a8 8 0 0 0 8-8V72a8 8 0 0 0-8-8zM171.2 38.4A16.1 16.1 0 0 1 184 32h80a16.1 16.1 0 0 1 12.8 6.4L296 64H152zM384 464a16 16 0 0 1-16 16H80a16 16 0 0 1-16-16V96h320zm-168-32h16a8 8 0 0 0 8-8V152a8 8 0 0 0-8-8h-16a8 8 0 0 0-8 8v272a8 8 0 0 0 8 8z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="makeStyles-container-9"
      >
        <p
          class="MuiTypography-root makeStyles-label-10 MuiTypography-body1"
        >
          Follow up on sales tips
        </p>
        <div
          class="makeStyles-actionElement-11"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButtonSize-12"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-trash-alt fa-w-14 "
                data-icon="trash-alt"
                data-prefix="fal"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M296 432h16a8 8 0 0 0 8-8V152a8 8 0 0 0-8-8h-16a8 8 0 0 0-8 8v272a8 8 0 0 0 8 8zm-160 0h16a8 8 0 0 0 8-8V152a8 8 0 0 0-8-8h-16a8 8 0 0 0-8 8v272a8 8 0 0 0 8 8zM440 64H336l-33.6-44.8A48 48 0 0 0 264 0h-80a48 48 0 0 0-38.4 19.2L112 64H8a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h24v368a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V96h24a8 8 0 0 0 8-8V72a8 8 0 0 0-8-8zM171.2 38.4A16.1 16.1 0 0 1 184 32h80a16.1 16.1 0 0 1 12.8 6.4L296 64H152zM384 464a16 16 0 0 1-16 16H80a16 16 0 0 1-16-16V96h320zm-168-32h16a8 8 0 0 0 8-8V152a8 8 0 0 0-8-8h-16a8 8 0 0 0-8 8v272a8 8 0 0 0 8 8z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
